signcrypt_seal.go

```diff
@@ -0,0 +1,262 @@
+
+	detachedSig, err := sss.signingKey.Sign(signatureInput)
+	if err != nil {
+		return err
+	}
+
+	attachedSig := append(detachedSig, b...)
+
+	ciphertext := secretbox.Seal([]byte{}, attachedSig, (*[24]byte)(nonce), (*[32]byte)(&sss.encryptionKey))
+
+	block := []interface{}{ciphertext}
+
+	if err := sss.encoder.Encode(block); err != nil {
+		return err
+	}
+
+	sss.numBlocks++
+	return nil
+}
+
+func receiverEntryForBoxKey(receiverBoxKey BoxPublicKey, ephemeralPriv BoxSecretKey, payloadKey SymmetricKey, index uint64) receiverKeys {
+	// Derive a shared secret by encrypting zeroes, as in the encryption
+	// format. Multiple recipients could potentially claim the same public key
+	// and share this secret.
+	sharedSecretBox := ephemeralPriv.Box(receiverBoxKey, nonceForDerivedSharedKey(), make([]byte, 32))
+	derivedKey, err := rawBoxKeyFromSlice(sharedSecretBox[len(sharedSecretBox)-32 : len(sharedSecretBox)])
+	if err != nil {
+		panic(err) // should be statically impossible, if the slice above is the right length
+	}
+
+	// Compute the identifier that the receiver will use to find this entry.
+	// Include the recipient index, so that this identifier is unique even if
+	// two recipients claim the same public key (though unfortunately that
+	// means that recipients will need to recompute the identifier for each
+	// entry in the recipients list). This identifier is somewhat redundant,
+	// because a recipient could instead just attempt to decrypt the payload
+	// key secretbox and see if it works, but including them adds a bit to
+	// anonymity by making box key recipients indistinguishable from symmetric
+	// key recipients.
+	keyIdentifierDigest := sha512.New()
+	keyIdentifierDigest.Write([]byte("saltpack signcryption derived key identifier\x00"))
+	keyIdentifierDigest.Write(derivedKey[:])
+	keyIdentifierDigest.Write(nonceForPayloadKeyBoxV2(index)[:])
+	identifier := keyIdentifierDigest.Sum(nil)[0:32]
+
+	payloadKeyBox := secretbox.Seal(
+		nil,
+		payloadKey[:],
+		(*[24]byte)(nonceForPayloadKeyBoxV2(index)),
+		(*[32]byte)(derivedKey))
+
+	return receiverKeys{
+		ReceiverKID:   identifier,
+		PayloadKeyBox: payloadKeyBox,
+	}
+}
+
+type ReceiverSymmetricKey struct {
+	// In practice these identifiers will be KBFS TLF keys.
+	Key SymmetricKey
+	// In practice these identifiers will be KBFS TLF pseudonyms.
+	Identifier []byte
+}
+
+func receiverEntryForSymmetricKey(receiverSymmetricKey ReceiverSymmetricKey, ephemeralPub BoxPublicKey, payloadKey SymmetricKey, index uint64) receiverKeys {
+	// Derive a message-specific shared secret by hashing the symmetric key and
+	// the ephemeral public key together. This lets us use nonces that are
+	// simple counters.
+	derivedKeyDigest := sha512.New()
+	derivedKeyDigest.Write([]byte(signcryptionSymmetricKeyContext))
+	derivedKeyDigest.Write(ephemeralPub.ToKID())
+	derivedKeyDigest.Write(receiverSymmetricKey.Key[:])
+	derivedKey, err := rawBoxKeyFromSlice(derivedKeyDigest.Sum(nil)[0:32])
```

Comment:
@mlsteele and I will probably want to change the way we mix context with keys here, based on what we've done in chat. I'll factor it out into a shared function too, since open duplicates this.